core/src/main/java/org/apache/accumulo/core/fate/zookeeper/ZooCache.java

```diff
@@ -207,8 +172,23 @@ public void process(WatchedEvent event) {
    */
   public ZooCache(ZooReader reader, Watcher watcher) {
     this.zReader = reader;
-    nodeCache = new ConcurrentHashMap<>();
     this.externalWatcher = watcher;
+    RemovalListener<String,ZcNode> removalListerner = (path, zcNode, reason) -> {
+      try {
+        log.trace("{} removing watches for {} because {}", cacheId, path, reason);
+        reader.getZooKeeper().removeWatches(path, watcher, Watcher.WatcherType.Any, false);
+      } catch (InterruptedException | KeeperException | RuntimeException e) {
+        log.warn("Failed to remove watches on path {} in zookeeper", path, e);
+      }
+    };
+    // Must register the removal listener using evictionListener inorder for removal to be mutually
+    // exclusive with any other operations on the same path. This is important for watcher
+    // consistency, concurrently adding and
+    // removing watches for the same path would leave zoocache in a really bad state. The cache
+    // builder has another way to register a removal listener that is not mutually exclusive.
+    Cache<String,ZcNode> cache = Caffeine.newBuilder().expireAfterAccess(3, TimeUnit.MINUTES)
```

Comment:
There is a `Caches` class that we are using to provide a common way to create Caffeine caches.

Reply:
Updated to use that in 5c65c4f